src/Workspaces/Core/Portable/Storage/CloudCache/ICloudCacheService.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Storage
+{
+    /// <summary>
+    /// Clone of the platform's Microsoft.VisualStudio.RpcContracts.Caching.ICacheService, but defined so we can
+    /// reference it in the workspace layer, and function if it is not present.
+    /// </summary>
+    internal interface ICloudCacheService : IDisposable
+    {
+        Task<bool> CheckExistsAsync(CloudCacheItemKey key, CancellationToken cancellationToken);
+        ValueTask<string> GetRelativePathBaseAsync(CancellationToken cancellationToken);
+        Task SetItemAsync(CloudCacheItemKey key, PipeReader reader, bool shareable, CancellationToken cancellationToken);
+        Task<bool> TryGetItemAsync(CloudCacheItemKey key, PipeWriter writer, CancellationToken cancellationToken);
+    }
+
+    internal readonly struct CloudCacheItemKey
+    {
+        public readonly CloudCacheContainerKey ContainerKey;
+        public readonly string ItemName;
+        public readonly ReadOnlyMemory<byte> Version;
+
+        public CloudCacheItemKey(CloudCacheContainerKey containerKey, string itemName, ReadOnlyMemory<byte> version = default)
+        {
+            ContainerKey = containerKey;
+            ItemName = itemName;
+            Version = version;
+        }
+    }
+
+    internal readonly struct CloudCacheContainerKey
+    {
+        public readonly string Component;
+        public readonly ImmutableSortedDictionary<string, string?> Dimensions;
+
+        public CloudCacheContainerKey(string component, ImmutableSortedDictionary<string, string?>? dimensions = null)
+        {
+            Component = component;
+            Dimensions = dimensions ?? ImmutableSortedDictionary<string, string?>.Empty;
```

Comment:
You should always set the comparer explicitly to `StringComparer.Ordional` so that it matches [what RPC.Contracts uses](https://dev.azure.com/devdiv/DevDiv/_git/VS.RPC.Contracts?path=%2Fsrc%2FMicrosoft.VisualStudio.RpcContracts%2FCaching%2FCacheContainerKey.cs&version=GBdev%2Fpafenelo%2FaddLogging_v0.1&line=36&lineEnd=36&lineStartColumn=84&lineEndColumn=128&lineStyle=plain&_a=contents) otherwise [a value-copy will be made](https://dev.azure.com/devdiv/DevDiv/_git/VS.RPC.Contracts?path=%2Ftest%2FMicrosoft.VisualStudio.RpcContracts.Tests%2FCaching%2FCacheContainerKeyTests.cs&version=GCfcb7d3102b6f6bced2424a503affbb2134d25205&line=93&lineEnd=103&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) when you eventually hand it off to that type.
The reason for this is we need deterministic ordering of the collection so that the hash of its content will also be deterministic.

Reply:
i'll do that. note though that yuou can likely avoid that check if the dictionary is empty, right?

Reply:
done.